src/langchain_google_spanner/chat_message_history.py

```diff
@@ -0,0 +1,128 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Cloud Spanner-based chat message history"""
+from __future__ import annotations
+
+import uuid
+from typing import List, Optional
+
+from google.cloud import spanner
+from google.cloud.spanner_admin_database_v1.types import DatabaseDialect  # type: ignore
+from google.cloud.spanner_v1 import param_types
+from google.cloud.spanner_v1.data_types import JsonObject
+from langchain_core.chat_history import BaseChatMessageHistory
+from langchain_core.messages import BaseMessage, message_to_dict, messages_from_dict
+
+OPERATION_TIMEOUT_SECONDS = 240
+
+COLUMN_FAMILY = "langchain"
+COLUMN_NAME = "history"
+
+
+class SpannerChatMessageHistory(BaseChatMessageHistory):
+    """Chat message history that stores history in Spanner.
+
+    Args:
+        instance_id: The Spanner instance to use for chat message history.
+        database_id: The Spanner database to use for chat message history.
+        table_name: The Spanner table to use for chat message history.
+        session_id: Optional. The existing session ID.
+    """
+
+    def __init__(
+        self,
+        instance_id: str,
+        database_id: str,
+        table_name: str = "message_store",
+        session_id: Optional[str] = None,
+        client: Optional[spanner.Client] = None,
+    ) -> None:
+        self.instance_id = instance_id
+        self.database_id = database_id
+        self.table_name = table_name
+        self.client = client or spanner.Client()
+        self.instance = self.client.instance(instance_id)
+        if not self.instance.exists():
+            raise Exception("Instance doesn't exist.")
+        self.database = self.instance.database(database_id)
+        if not self.database.exists():
+            raise Exception("Database doesn't exist.")
+        self.database.reload()
+        self.dialect = self.database.database_dialect
+        self.session_id = session_id or uuid.uuid4().hex
+        self._create_table_if_not_exists()
+
+    def _create_table_if_not_exists(self) -> None:
+        google_schema = f"""CREATE TABLE IF NOT EXISTS {self.table_name} (
+                            id STRING(36) DEFAULT (GENERATE_UUID()),
+                            created_at TIMESTAMP NOT NULL OPTIONS (allow_commit_timestamp=true),
+                            session_id STRING(MAX) NOT NULL,
+                            message JSON NOT NULL,
+                         ) PRIMARY KEY (session_id, created_at ASC, id)"""
+
+        pg_schema = f"""CREATE TABLE IF NOT EXISTS {self.table_name}  (
+                             id varchar(36) DEFAULT (spanner.generate_uuid()),
+                             created_at SPANNER.COMMIT_TIMESTAMP NOT NULL,
+                             session_id TEXT NOT NULL,
+                             message JSONB NOT NULL,
+                             PRIMARY KEY (session_id, created_at ASC, id)
```

Comment:
Remove "ASC"? I am getting Error: google.api_core.exceptions.InvalidArgument: 400 [ERROR] syntax error at or near "ASC"; failed to parse the DDL statements.